### PR TITLE
Add hallucination filter

### PR DESCRIPTION
**Pull request by KMFODA**
# What does this PR do?

As per the discussions in https://github.com/huggingface/transformers/issues/18354, this PR aims to add a `HallucinationPenaltyLogitsProcessor` that takes in a `hallucination_penalty` that is applied to any tokens that are not in the original input. This acts as a hallucination filter and the higher it is set the more likely the text is going to contain only the input tokens. For summarisation this means the higher the hallucination penalty the more extractive the summary is going to be and the less likely it is to have a hallucination.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@gante, @patrickvonplaten 

Library:

- text generation: @patrickvonplaten

### CODE REVIEW

File: src/transformers/generation_logits_process.py
Comment (patrickvonplaten) [inline src/transformers/generation_logits_process.py:170]: Generally this looks good to me, but I'd be in favor to write the class more in the style of https://github.com/huggingface/transformers/blob/06d1ba1a55a12b3fb3ca081bdd4f812fda800c37/src/transformers/generation_logits_process.py#L144 so that users only have to know one way of applying the penalty - then we could also just call it `EncoderRepetitionPenalty...` which is IMO a bit more descriptive than `Hallucination....`

File: src/transformers/generation_logits_process.py
Comment (patrickvonplaten) [inline src/transformers/generation_logits_process.py:190]:
```suggestion
        score = torch.where(score < 0, score * self.penalty, score / self.penalty)
```
-> can we write it this way and make it a "penalty" logits processor so that it's 1-to-1 aligned with the one we already have? 
Then we can just pass `1 / penalty` instead of `penalty` for your use case :-)

**Comment by gante** [comment]
Hi @KMFODA are you still planning to work on this? We can reopen the PR :)

**Comment by KMFODA** [comment]
Hey @gante yes I still plan to work on this. My apologies I had fallen ill this past month and couldn't spend time on this. If you re-open the PR I will prioritise working on this early next month.

**Comment by gante** [comment]
@KMFODA absolutely no rush, take your time -- personal health always takes precedence! I hope you're feeling better 🤗

File: src/transformers/generation_logits_process.py
Comment (KMFODA) [reply src/transformers/generation_logits_process.py]: Sure thing, I just changed the penalty and the name to `EncoderRepetitionPenaltyLogitsProcessor`. Let me know how that looks now :)

**Comment by gante** [comment]
(@KMFODA let us know when you'd like a new review)

**Review by gante — COMMENTED** [review]
Other than a few minor nits and a pair of missing tests (a numerical like [this one](https://github.com/huggingface/transformers/blob/main/tests/generation/test_generation_logits_process.py#L106) and an integration test like [this one](https://github.com/huggingface/transformers/blob/0e83c9664b2a440ade59066a77fb01d0143e4d18/tests/generation/test_generation_utils.py#L1663) where the new processor is tested in practice), seems great 🙌

**Comment by KMFODA** [comment]
Hi @gante, I added a `test_encoder_repetition_penalty_dist_process` to covert the 1st type of test. The 2nd test you've linked seems to be more focused on beam searches and stopping criteria. What type of test did you have in mind here for the encoder_repetition_penalty? Would ensuring it's initialised by adding it to this [test](https://github.com/huggingface/transformers/blob/0e83c9664b2a440ade59066a77fb01d0143e4d18/tests/generation/test_generation_utils.py#L101) cover this?

File: tests/generation/test_logits_process.py
Comment (gante) [inline tests/generation/test_logits_process.py:148]: We should also check that the other values were not touched, e.g.:
```suggestion
        # check that values not in the encoder ids were NOT changed
        self.assertAlmostEqual(scores[0, 2].item(), (1 / vocab_size))
        self.assertAlmostEqual(scores[1, 2].item(), (1 / vocab_size))
```

**Review by gante — COMMENTED** [review]
Two minor nits and should be ready to go (assuming tests are passing) :)

**Review by gante — APPROVED** [review]
LGTM!

This can be converted from draft to a PR, to then be merged.

cc @patrickvonplaten for a 2nd set of eyes

**Comment by KMFODA** [comment]
Done, thanks for all the helpful comments to get this merged and apologies it took so long.

**Comment by gante** [comment]
@KMFODA did this processor help with your use case? :)

File: src/transformers/generation/logits_process.py
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:176]: Missing documentation for the `encoder_input_ids` 😉

**Review by ArthurZucker — APPROVED** [review]
Hey! Thanks for the addition! On my side just a few docstring missing.

**Comment by KMFODA** [comment]
Thanks @ArthurZucker. Added the doc strings in the 3 different files you mentioned. I've only got one test failing which I can't recreate locally:

`tests/pipelines/test_pipelines_zero_shot.py::ZeroShotClassificationPipelineTests::test_small_model_tf`

It seems like the outputs changed in the `zero-shot-classification` pipeline although I'm not sure why. Are you able to point me towards what might be causing this to fail?

**Comment by KMFODA** [comment]
Managed to fix the failing test by rebasing to main. Hopefully should be good to merge now but if not let me know!

**Comment by ArthurZucker** [comment]
Cool let's just ask for a final review from @sgugger ! 🤗

**Review by sgugger — COMMENTED** [review]
The PR looks good to me in general, but it would be awesome if we can wait for #20388 to be merged and only add this new argument in the `GenerationConfig`. Since @gante is back from vacation it shouldn't be too long now.

**Comment by gante** [comment]
Hey @KMFODA -- the big change we just merged clashed with your PR, as @sgugger mentioned above.

In a nutshell, new generation parameters should go in `GenerationConfig` ([here](https://github.com/huggingface/transformers/blob/1543cee7c8c95ef47f832b1f37625ba2923c4994/src/transformers/generation/configuration_utils.py#L38)), and generate always uses a generation config (related [docs](https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)). We need to change this PR to make it consistent with the new changes :D

I understand this PR has been a long process, so it's okay if you don't want to make further changes. Just let me know if you are no longer interested in working on it :)

**Comment by KMFODA** [comment]
Hey @gante thanks for getting back to me. Not a problem, I'd still like to work on this as it'll help me learn about the new generation engine. I'll get working on this after the holidays.

**Comment by KMFODA** [comment]
Hey @gante, moved encoder_repetition_penalty to `GenerationConfig` and fixed all failing tests. Let me know if more is needed to merge this PR.

File: src/transformers/generation/configuration_utils.py
Comment (gante) [inline src/transformers/generation/configuration_utils.py:118]:
```suggestion
        encoder_repetition_penalty (`float`, *optional*, defaults to 1.0):
```

**Review by gante — APPROVED** [review]
Two minor changes (added as suggested changes) and it's ready to go 👍

**Comment by KMFODA** [comment]
Thanks @gante helpful changes. Just implemented and fixed failing tests.

**Comment by gante** [comment]
@KMFODA awesome, thanks 🙏 

@sgugger this PR should be ready to go in, feel free to merge if you are also happy with it :)

File: src/transformers/generation/logits_process.py
Comment (sgugger) [inline src/transformers/generation/logits_process.py:216]: Nit the type is not the same as the type annotation below.

File: tests/test_configuration_common.py
Comment (sgugger) [inline tests/test_configuration_common.py:335]: There is no reason to reformat this here. Please revert this change.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! Can you just make sure to remove the styling changes not related to this PR? Then we can merge this :-)

File: tests/test_configuration_common.py
Comment (KMFODA) [reply tests/test_configuration_common.py]: Thanks for the comments @sgugger. I fixed the first. This specific reformat was enforced when running `make style`. If I revert it the code fails the quality check.

**Comment by KMFODA** [comment]
Hopefully should be good to merge now. If not let me know.

File: tests/test_configuration_common.py
Comment (gante) [reply tests/test_configuration_common.py]: @KMFODA it can be fixed by removing the trailing comma in `"transformers_version",` and then running `make style` :) Alternatively, you can copy/paste from the original.

**Comment by sgugger** [comment]
Thanks for your contribution!

**Pull request merged**
